### DISCUSSION TITLE
Option to synchronize selection across all visible views

### DISCUSSION THREAD

**Discussion by aaa1386** [freeplane/freeplane#1936] [Ideas]
I propose a very practical and excellent idea, I hope you will pay attention to it.

First, you should familiarize yourself with the terms that I defined myself. Well, pay attention, I explain the idea according to these terms.
🟢 Terms:
-Imagine a person named Ali, he has two brothers named Hamid and Muhammad. He has three uncles. He also has three children. Look at the image below. When I say in my explanation, the column of fathers and uncles, or the column of Ali and brothers, or the column of children, you can see these columns respectively on the map.

![image](https://github.com/user-attachments/assets/e36fa6c1-352c-4f9b-b4bf-6ba3f965ef48)

2-Three orange arrows show the color of the columns or levels. Pay attention that level 3 is lower than level 1. Because it is farther from the root.
3-  Look at the image below: By implementing the idea, you will have twin panels, which I call the main and auxiliary or secondary panels. The main panel is exactly like the FP maps, and the secondary or auxiliary panel is different and has a different function.

Even now, two views of the map can be seen side by side, but there is no connection between these two views, like two bricks side by side that have no connection with each other, they are just placed next to each other.
By implementing this idea, many connections are established between two views of the map.

![image](https://github.com/user-attachments/assets/eeb67996-becc-422d-b11f-b1c9826c02a9)

🟢🟢 Why this idea came up:
I wrote some questions in FP. All questions were in one column or level. The answer to all those questions had many branches and nodes.
 When I checked the answer nodes, I couldn't see any questions that were all level 1. I was thinking that I wish it was possible to see those questions in a separate  and I could easily select that particular question whenever I wanted. . Similar to bookmarks in a PDF file, you can always view the entire level of bookmarks in the side panel.

another point:
One of the meanings of the mind map is to be able to obtain information about the structure and framework of a number of items and their relationship with each other in one moment.
I give an example:
Look at the image below:
![image](https://github.com/user-attachments/assets/3bea13fd-4cf6-4aca-8927-42c40d57ca26)

Imagine the "colors" node.
 In the map, three things are important for concentration: simultaneously seeing the nodes of brothers, the nodes of fathers and uncles, and the nodes of children.
I drew a line around these things.

I will explain more later, but if the "Ali" node is selected:
 The uncles or brothers or both of them are displayed in the auxiliary panel depending on the modes set by the user.

Next point:

Suppose we have a node named "A".
The problem is that if the map has many nodes and branches, or if all the brothers of "A" are unfolded, or if node "A" has many children (for example, 30 children), in these cases, the brothers of node "A" are not seen. 

Another problem is that the uncles of the selected node are not seen on the screen at the same time in many cases. Even the father of that node is not obvious in many cases if the `Scroll map after selecting a node` setting is active.

 This idea can have many side benefits in addition to solving the above problem. To get to know the idea briefly, imagine the bookmarks panel of a PDF display software and the navigation panel of Windows Explorer and Microsoft Word. Imagine how helpful these tools can be and create a lot of mental focus for the user.
We are looking to create a panel that can do very interesting things.

🟢🟢  Some benefits of this idea:
• Ability to see father, uncles and brothers at the same time.
• More mental focus on the map - like the help panels in some of the mentioned software.
• Ease of copying and pasting nodes in the map with minimal folding or unfolding of nodes.
• If the distance between the brothers is large, copying and pasting is difficult, but with this idea it is very easy to do.

• Very easy navigation on the map.

• New navigation mode in the auxiliary panel among layers and columns using the four buttons for the direction of the view.

• Improve the efficiency of the `Scroll map after selecting a node`  option. This option takes the selected node to the beginning of the screen. (In this case, the father and uncle of that node are not displayed, but the father of that node can be viewed using the auxiliary panel.

• When you select a node in the auxiliary panel, it causes an automatic scrolling to occur in the main panel. It means that you have selected the desired node and set the map so that the most nodes can be seen.

• Prevents many opening and closing. You don't have to worry about how many nodes of the map are open or closed. The side panel helps you.

🟢🟢 How to create two panels?

FP has a feature that you can display two maps on the screen at the same time or display one map in such a way that half of the screen shows that map and the other half of the screen shows the same map.

 The way to do it is to mount the idea on one of these two panels. Whenever we implement the option of twin panels, our idea will be implemented, but the basis of the work is to display two maps at the same time in FP. In fact, it is two real views of the map that have an effect on each other. And each has its own characteristics.

🟢🟢Rules of twin panels:

1- Four different modes of displaying nodes in the auxiliary panel: 1- Show only brothers 2- Show father and uncles 3- Show brothers and father and uncles 4- Lock on the selection column.
 You can tell FP to lock on that column in the auxiliary panel by dragging a node from the selected column or layer to the mode 4 icon.

• Relationship between two panels:

•  Whenever we click on a node in the main panel, for example, if we have activated the  mode 2, the father and uncle of that node will be displayed in the auxiliary panel. 

•Whenever another node is selected in the main panel, the necessary changes to display fathers, uncles, etc. are updated in the auxiliary panel.

•   If one of the uncles is clicked in the auxiliary panel, his children will be displayed in the main panel.
•  etc.

• Sometimes the nodes are very long, so the user has a setting that allows him to choose two modes: display 5 words from each node or fully display the nodes (if the number of nodes is small, the full display will be automatic)

 The auxiliary panel does not always show more than one level or at most two levels (when the display mode of brothers, uncles and father is active).
 By selecting a node from the sub-map, all its children will appear in the main panel. The map should appear in the main panel in such a way that the most nodes are displayed (i.e. automatic scrolling).
 If mode 1 is selected, i.e. "only brothers" display mode, in the auxiliary panel, in addition to the display of brothers, the father of the node selected in the main panel will be shown in a separate color at the top of all brothers.

• Higher or lower levels are shown in the auxiliary panel with the right and left arrow keys. In this case, the highest node of the displayed level must be defined by FP to display its children in the main panel.
This is a new navigation mode in auxiliary panel. If you imagine it well, this navigation with right and left and up and down arrow keys is very useful.

**Comment by aaa1386** [comment]
Microsoft Word software, PDF file viewers, Windows Explorer, etc. have created a panel for more focus.
If the idea expressed in FP is implemented, we will benefit from the advantages of the mentioned software, but many more benefits will be created in FP.

**Comment by aaa1386** [comment]
A point to better understand the necessity of this idea.
Suppose you have made a PDF with ten bookmarks in one level.
PDF viewer software has an interesting function that is very logical.
They have a panel that is always visible to the user for bookmarks and a main panel for displaying the selected bookmark text.
It doesn't matter how much the contents of a bookmark are organized and charted or not? The important thing is that you have ten bookmarks that you always have to scroll through so it should always be visible to you.
Although an FP map is made up of many nodes, this logic exists in FP maps, between nodes and branches.
 Sometimes a level or column like those ten bookmarks that we said should always be visible to the user. This level or column of nodes has children and branches. Consider all these children and branches as the text of that bookmark. In the idea that was expressed, the auxiliary panel works like the bookmarks panel in the software that shows the PDF file. Of course, this idea is much more flexible and attractive than other software.

The existence of this idea makes the user unconsciously encouraged and forced to create a clearer logical connection between the nodes.

Even if sometimes there is no clear logical connection between the auxiliary panel and the main panel, the least effect of this auxiliary panel is in easy navigation between maps and moving nodes between maps.

In my opinion, the existence of this idea makes the FP map have a very beautiful and attractive logic.

**Comment by aaa1386** [comment]
another point :
By implementing this idea, nothing is lost from the environment that shows the map, FP works exactly as before in the main map. This idea just adds an auxiliary panel to the map, which is very useful.

**Comment by aaa1386** [comment]
Another point:
One of the important applications is to solve the problem of long distance between nodes, which exists in many cases.
We know that many nodes have a logical connection and dependence on each other, that's why we should always observe them simultaneously. But the long distance between these nodes makes us unable to observe them at the same time.
 I have shown the presence of a large distance in the image below.
This idea solves this problem to a large extent. Especially since the auxiliary panel shows the items that usually have the most and closest logical connection to each other. Like brothers, uncles, or the user's favorite column.
 Although folding the nodes is a solution to view the nodes at the same time, but in many cases it is not the best solution because you have to move from your node and branch to the root and get closer to be able to close the nodes. Sometimes you don't want to lose the extent of the map while you also need to see some nodes together.

![image](https://github.com/user-attachments/assets/28bdd55a-5cce-4abf-885c-bd076d72554e)

**Comment by aaa1386** [comment]
If it is possible to create an auto-hide feature for the auxiliary panel, it will be great.

**Comment by euu2021** [comment]
I suggest that you open a discussion. The topic is too complex for a feature request ticket.

For example, I understand the "problem of long distance between nodes", but I don't understand this situation:

> Why this idea came up:
I wrote some questions in FP. All questions were in one column or level. The answer to all those questions had many branches and nodes.
When I checked the answer nodes, I couldn't see any questions that were all level 1. I was thinking that I wish it was possible to see those questions in a separate and I could easily select that particular question whenever I wanted. . Similar to bookmarks in a PDF file, you can always view the entire level of bookmarks in the side panel.

**Comment by dpolivaev** [comment]
I want to try something here.

**Comment by dpolivaev** [comment]
preview 1.12.2_11 alpha
View->New map view
fold/unfold all nodes on the second view and arrange views
activate View->View settings->Synchronize selection

**Reply by euu2021** [reply]
This is really cool. It's like a Two-pane outliner. And it also works with more than two panes, so it's an Multiple-pane outliner.

**Comment by aaa1386** [comment]
> preview 1.12.2_11 

Thank you for your attention to this idea.
But the "Synchronize selection" option implements a small amount of the expressed idea.
In addition, it has many problems, especially sometimes in different folding or unfolding modes, it does not work in the view that is activated and it works in the opposite view that is not active.



https://github.com/user-attachments/assets/ee0b8a58-6f17-4588-907d-513cacb717a1


https://github.com/user-attachments/assets/73376977-f83f-4c80-b953-2935e1c0a018


https://github.com/user-attachments/assets/c9bdb4e0-81e5-43c9-ba04-328f3050495e

There are many such errors.
Currently, despite these errors in folding or unfolding in the active view, this option cannot be used.

**Comment by dpolivaev** [comment]
In all recent video examples I see no errors. Could you please explain what you think is wrong?

**Comment by dpolivaev** [comment]
Generally, Freeplane does not make any difference between full panel and auxiliary panel, it only makes difference between currently selected panel and other visible panels where the selection is adjusted.

**Comment by dpolivaev** [comment]
I am confused, but I see no errors.

**Reply by aaa1386** [reply]
https://github.com/user-attachments/assets/f835945f-4a2a-4f69-9a9b-8540c11becef




I want the "aa" panel nodes to always stay folded. But I can fold or unfold the nodes of the "BB" panel. I unfold nodes 1, 2, and 3 respectively in the "BB" panel. Node "2" is unfolded in the "AA" panel. it's wrong.

**Reply by dpolivaev** [reply]
Thanks, now I see it.

**Comment by dpolivaev** [comment]
Preview 1.12.2_12 alpha

**Reply by aaa1386** [reply]
Pay attention to the movement of the mouse in selecting the nodes. The left panel works correctly, but when I select node "1" and the root, the right panel does not react.

https://github.com/user-attachments/assets/f560da6a-aec5-412d-b88a-36fdb60cee02

**Reply by dpolivaev** [reply]
If you select node "1" on the right, node "1" does not become selected on the left, if any its subnode e.g. 1->1  is already selected. I intentionally implemented it this way. If you do not like it, explain your use case better and explain why current solution does not satisfy it.

**Reply by aaa1386** [reply]
I don't know what your purpose was. If you can explain maybe I can check better. 

So let us work for a while and then I will give my opinion. Of course, it may take some time.

**Reply by aaa1386** [reply]
![image](https://github.com/user-attachments/assets/8d996c34-428d-48d0-8c1d-9524e4537500)
Look at this picture.
I have opened two views of the same map. I write this comment after thinking a lot.

Current function of FP: suppose I select node 1->1 with the mouse in the left view.
Then I go to the right view and if I  select select sub-nodes of node 1->1 with the mouse, node 1->1 remains active and selected on the right view . So far everything is fine,
 but when I select node 1 on the right  view ,    node 1 on the left side is not activated. 
This problem becomes more complicated when if we select node 2 in the right view and then select node 1, node 1 will be activated in the left view. That is, node 1 on the left is sometimes activated and sometimes not

Although it is difficult to express what I have in mind, I will try to explain.

 1- The current function of FP confuses users who use the "synchronize selection" option for the first time. They may give up the option altogether.

2- I did not find any use for the current function of FP.

2- The operation of this option should be simple and transparent, that is, a simple rule that is similar to the operation of the bookmark panel in software that displays PDF files.
 Whenever the user selects a node using the mouse or arrow keys in one view, FP must automatically select the same node in the other view, or if that node is not obvious, it must select its father or ancestor.

 This simple rule makes sure that when the user selects a node in the other view, the node that is automatically selected is the exact same node or is the parent or ancestor or higher ancestors.

3- If the above simple rule is implemented, it can be useful for the user. Suppose the user has created nodes in different places of the map with "Structure" style. In the left view, it filters the "Structure" style, so in the left view, nodes that have this style are always displayed. The idea I proposed is very suitable for the user to focus on the map.

4- Suppose the user shrinks the left view to the length of a node, so a column is always displayed in the left view, when the user selects a node in the right view, its brothers are also displayed in the left view، especially if all the nodes of that column are folded.

5- In my opinion, other benefits can be imagined for it.


I suggest you change the function of FP according to the idea that has been proposed - thank you very much for the good FP software.
🍀🍀🍀✅✅

**Comment by aaa1386** [comment]
Now one of the modes of the idea has been created.

Is the following possible?
 When "Synchronize selection" mode is executed. FP will open two views of the map.

Let's define several modes for "Synchronize selection":
1- Display mode of brothers in view "1"
2- Display mode of father and uncles in view "1"
3- Display mode of brothers, father and uncles in view "1"
4- Display mode of brothers, father, uncles and children (meaning the first generation, not the second generation and above.)
5-Free states (this means the state that is currently implemented)

Suppose we have chosen the first mode. When we click on any node in view "2", only all brothers of that node will be displayed in view "1". If we select another node, its brothers will be displayed in the "1" view.

Be sure that these five modes are very practical. I tried myself. The decision is with the creators of FP.


I will explain the reason for this idea with an example. Of course, the benefits of this idea are not only related to situations like this example, it has many benefits.
Suppose you have a category. You have divided apples, oranges, cucumbers according to size, color and taste.

Imagine the color of apples. There are several important things to see at the same time:
 Know that you are now checking the color of the apple (i.e. the father in the FP map).
 Know that the three colors of green, yellow, red are its branches.
 Know that in addition to checking the color of the apple, you also want to check the taste and size. (Brothers on the FP map)
Know that you are now checking the color of apples, and in addition to that, you want to check for oranges and cucumbers (uncles in the FP map).

**Reply by aaa1386** [reply]
Please let me know if it is possible or not.



If this idea is implemented with these 5 modes. I think many people always use two views to execute the map.

**Reply by aaa1386** [reply]
> Now one of the modes of the idea has been created.
> 
> Is the following possible?
...

Mode 6 can also be added.
This mode says lock on the desired column or level in view "1".
In this case, only the desired column that is locked will be displayed in all cases.

**Reply by aaa1386** [reply]
If this idea is implemented with various modes that have "1" view. The user can easily keep the "2" view almost always fully unfolded.

**Reply by dpolivaev** [reply]
I believe that 5 modes are 4 modes too much.

**Reply by aaa1386** [reply]
Yes, it may be so.

**Reply by aaa1386** [reply]
Of course, there is a possibility that one of these modes will be useful for the user at any time, and the existence of all of them will make the use of the idea more complete. I think it's good to have more modes to start with, over time users will judge which one is less useful.
In my opinion, what is important and should be considered is easy access to these modes so that the user can quickly switch between them.
This was my opinion, you may have another opinion.

**Comment by dpolivaev** [comment]
I moved the option in menu "Navigate" in  preview 1.12.2_13

**Comment by euu2021** [comment]
I've been trying this new feature and I really liked it. It helps a lot when working on big branches, because it's possible to keep a view with just the main structure, to be used for quick navigation. 

Thanks for the feature request, ajaberi1360.

**Reply by euu2021** [reply]
And thanks to Dimitry, for implementing it

**Reply by aaa1386** [reply]
> I've been trying this new feature and I really liked it. 

Thank you for your comment:
I am waiting to see if this idea is finished or something else is added to it.
After I am sure, I will come up with another very beautiful idea. Of course, before that idea, I'd like to open a page to solve some FP problems with right-to-left based languages ​​if possible.
